[AN] Refactor: 홈 화면 uiState 분리, StateFlow 도입

HamBeomJoon opened the pull request:
## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->
- viewModel에 livedata 변수가 너무 많음
- homeUiState로 uiState + 로딩 상태 관리
- LiveData -> StateFlow로 변환

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->
- viewModel에 흩어져있던 상태변수들을 HomeUiState로 묶어서 관리했습니다.
- 또한, 바꾸는 김에 LiveData -> StateFlow로 변경했습니다.
- 로딩 상태에 따른 visible에서도 dataBinding등도 다 없애고 fragment에서 처리하도록  변경했습니다.

- async/await으로 api 5개를 관리하고 있었는데, api호출 함수를 각각으로 분리하고 loading중인 api호출 작업과 finish된 작업을 key로 관리하도록 바꿨는데, 어떻게 생각하시나요??
```
private fun startLoading(jobKey: String) {
    loadingJobs.add(jobKey)
    _uiState.update { it.copy(isLoading = true) }
}

private fun finishLoading(jobKey: String) {
    loadingJobs.remove(jobKey)
    if (loadingJobs.isEmpty()) {
        _uiState.update { it.copy(isLoading = false) }
    }
}
```


## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
#711 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새로운 기능**
  * Wootaeco 항목 선택 시 분석 이벤트 로그 추가

* **리팩토링**
  * 홈 화면을 통합된 UI 상태로 전환해 섹션별(추천, 재생 기록, 최근 업로드, 북마크 등) 동기화된 업데이트 제공
  * 라이프사이클 기반 코루틴 수집으로 UI 업데이트 신뢰성 향상
  * 섹션별 개별 로딩 키 도입으로 로딩 표시 및 오류 처리 개선

* **문서/문구**
  * 여러 UI 문자열 문구·구두점 정리 및 리소스 키 일부명 변경 (홈 추천 카테고리 텍스트)
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HamBeomJoon committed:
[AN] feat: 우테코 카테고리 진입 analytics 로깅 추가 #711

HamBeomJoon committed:
[AN] feat: HomeUiState 추가 #711

HamBeomJoon committed:
[AN] refactor: HomeViewModel LiveData를 StateFlow로 마이그레이션 #711

*   [AN] refactor: HomeViewModel의 UI 상태를 StateFlow로 통합 관리
*   [AN] refactor: 병렬로 실행되던 데이터 로딩 로직을 개별 비동기 호출로 분리

HamBeomJoon committed:
[AN] refactor: HomeFragment 상태 관리를 StateFlow로 전환 #711

*   [AN] refactor: LiveData observe 로직을 repeatOnLifecycle 내에서 StateFlow를 collect하도록 변경
*   [AN] refactor: UI 상태 업데이트 로직을 StateFlow의 UiState 모델을 사용하도록 개선
*   [AN] refactor: 데이터 바인딩 ViewModel 직접 사용 제거

HamBeomJoon committed:
[AN] refactor: HomeFragment의 데이터 바인딩 로직 개선 #711

- HomeFragment 레이아웃에서 불필요한 `viewModel` 변수를 제거합니다.
- `visibleIfCondition` 속성을 제거하여 뷰의 가시성 제어 로직을 단순화합니다.

HamBeomJoon committed:
[AN] refactor: 홈 화면 UI 업데이트 로직 개선 #711

*   [AN] refactor: 홈 화면 섹션별 UI 업데이트 로직 분리
*   [AN] refactor: 데이터 로딩 상태에 따른 UI 업데이트 로직 개선
*   [AN] refactor: navigateClickListener 람다를 메소드 참조로 변경

HamBeomJoon committed:
[AN] refactor: showToast 함수 string대신 resId받도록 변경 #711

HamBeomJoon commented:
@coderabbitai review

rosemin928 commented on android/app/src/main/java/com/onair/hearit/presentation/home/HomeViewModel.kt:
```diff
@@ -29,93 +26,115 @@ class HomeViewModel(
-    private val _isLoading = MutableLiveData(false)
-    val isLoading: LiveData<Boolean> = _isLoading
+    // 로딩 중인 작업 개수 추적
+    private val loadingJobs = mutableSetOf<String>()
```
loading을 key로 관리하는 것 좋네요! 👍
몇 가지 제안하고 싶은 점이 있어서 코멘트 남깁니당

1. 로딩키를 String이 아닌 `enum class`를 써서 한정시켜보는 것은 어떨까요?
2. 로딩 상태를 HomeUiState 내부로 옮겨서 관리해보는 것은 어떨까요? 지금 상태에서 ViewModel 내부 Set과 UI의 isLoading이 불일치 하는 상태가 생길 여지는 없을까요??

HamBeomJoon replied:
1. 그냥 String으로 관리하는건 안좋아보이긴하네요. 생각해보니 companion object에 상수로 넣고 쓰는 방식도 있을 것 같은데, enum을 써서 '한정'시키려는 건 어떤 의도인가요? 뭔가 기능이 더 추가되면 enum 상수가 추가되는건 마찬가지일 거 같기도 한데요. 그냥 오타나 그런걸 방지하기 위해서인가요?

2.  UI의 isLoading이 어떤거죠?? +  이미 UiState에 있긴합니다
```
data class HomeUiState(
    val userInfo: UserInfo = UserInfo.default(),
    val isLoggedIn: Boolean = false,
    val recommendHearits: List<RecommendHearit> = emptyList(),
    val playingHistoryHearits: List<PlayingHistoryHearit> = emptyList(),
    val recentUploadHearits: List<RecentUploadHearit> = emptyList(),
    val playingBookmarkHearits: List<Bookmark> = emptyList(),
    val recommendationCategories: List<RecommendationCategories> = emptyList(),
    val isLoading: Boolean,
)
```

rosemin928 replied:
1. 우선 enum class로 바꾸는게 어떻냐고 제안을 드린 것은 우선 미플이 말씀하신대로 오타를 방지하기 위함도 맞고, 그에 따른 타입 안정성에 있어요! 또, 코드 안정성과 유지보수에 훨씬 유리할거라 생각을 해서 제안을 드렸습니다!

2. 
`UI의 로딩 = UiState.isLoading, 진짜 로딩 상태 = ViewModel 내부의 Set`으로, 예외/취소로 finishLoading이 누락되거나, 동일 키로 중복 호출을 해서 Set에 여전히 키가 남는다면 불일치가 일어날 수 있을 것 같다고 생각했습니다.
그래서 로딩 상태를 `단일 소스`로 합치면 어떨까 생각해서 드린 제안이었는데 제가 말을 애매하게 한 것 같네요 😅

제가 생각했던 코드의 느낌은 
```kotlin
data class HomeUiState(
    val loadingKeys: Set<LoadingKey> = emptySet(),
    // ...
) {
    val isLoading: Boolean
        get() = loadingKeys.isNotEmpty()   // 파생값
}
```

이런 느낌이었습니다! 이런 구조에 확장 함수를 사용하면 좀 더 깔끔하게 ViewModel이 정리될 수도 있을 것 같은데 미플은 어떻게 생각하시나요??

rosemin928 replied:
지금도 사실 충분히 괜찮다고 생각하니 고민만 해주셔도 됩니당

HamBeomJoon replied:
1. enum class로 분리했습니다
2. 오 좋네요👍 해당 방법으로 수정했습니다. +근데 Set은 그냥 중복제거? 그런거 때문에 쓴건가요?

rosemin928 replied:
Set은 중복제거 때문에 쓴 것이 맞습니당 👍

rosemin928 commented on android/app/src/main/java/com/onair/hearit/presentation/home/HomeViewModel.kt:
```diff
@@ -29,93 +26,115 @@ class HomeViewModel(
     }
 
     private fun fetchData() {
-        _isLoading.value = true
+        fetchRecommendHearits()
+        fetchPlayingHistory()
+        fetchRecentUpload()
+        fetchBookmarks()
+        fetchCategories()
+    }
```
만약 현 상황에서 하나라도 finishLoading이 취소/예외로 누락되면 isLoading이 영구적으로 true로 남지는 않나요...???

HamBeomJoon replied:
safeLoad 함수를 만들어 실패/성공 이후에 finally로 finish작업은 일단 수행하도록 했습니다
```
private suspend inline fun <T> safeLoad(
        jobKey: HomeLoadKey,
        block: () -> T,
    ): T? {
        startLoading(jobKey)
        return try {
            block()
        } finally {
            withContext(NonCancellable) { finishLoading(jobKey) }
        }
    }
```

HamBeomJoon replied:
https://github.com/user-attachments/assets/b386fd15-95fc-4b3d-b7a0-0739f267b6df

최근 업로드된 히어릿 실패한 경우 토스트메세지 출력 + 나머지 4개 api만 화면에 표시인데, 흠 UX가 고민이네요

rosemin928 replied:
이 부분은 저희 예전에 회의에서 나왔던 의견대로 로딩이 안 된 것은 로딩이 안 됐다는 것을 보여줄 수 있는 UI를 띄워주고, 당겨서 새로고침 기능을 넣어보는 것도 좋을 것 같네요!

rosemin928 commented on android/app/src/main/java/com/onair/hearit/presentation/home/HomeFragment.kt:
```diff
@@ -36,6 +44,7 @@ import com.onair.hearit.presentation.main.DrawerClickListener
 import com.onair.hearit.presentation.search.category.CategoryComposeFragment
+import kotlinx.coroutines.launch
 
 class HomeFragment :
```
추가적으로, 혹시 `home` 패키지 내의 파일들을 정리해보는건 어떨까요??
뭔가 안에 파일들이 엄청 많아서 어댑터나 뷰홀더와 같은 부분들은 따로 내부 패키지로 정리해줘도 좋을 것 같아요!

HamBeomJoon replied:
저도 정리하고 싶다 생각들었어요. 근데, 저도 adapter 패키지, viewHolder 패키지로 나눠볼까했는데 보통 기능을 수정한다고 생각할때 해당 adapter, viewHolder를 같이 보거나 같이 수정해야 하는데 얘네가 또 패키지가 따로 나눠져있으면 뭔가 찾는게 불편할거 같기도 해서 일단 놔뒀습니다 🥲 어떻게 나누는게 좋을까요??

rosemin928 replied:
음...아니면 adapter 안에 종류별로 나눠보는건 어떠신가요? home-adapter 패키지 안에 `recommend`, `recommendationCategory` 이런 식으로...? 근데 이러면 패키지명이 너무 길어질 것 같기도 하네요...내일 같이 고민해봅시다...

HamBeomJoon replied:
다른 팀한테 물어봤는데, 그냥 adapter라는 패키지에 adapter, viewHolder를 넣어놨다고 하더라고요 (지금은 다 compose로 바꿔서 요즘 누가 viewHolder 씀? 이러긴 하던데 ☠️)

rosemin928 commented on android/app/src/main/java/com/onair/hearit/presentation/home/HomeFragment.kt:
```diff
@@ -170,50 +172,94 @@ class HomeFragment :
+        playingBookmarkHearits: List<Bookmark>,
+        shouldShow: Boolean,
+    ) {
+        binding.tvHomePlayingBookmarkTitle.isVisible = shouldShow
+        binding.rvHomePlayingBookmark.isVisible = shouldShow
+        playingBookmarkAdapter.submitList(playingBookmarkHearits)
```
데이터바인딩을 전부 없애고 fragment에서 가시성을 관리하도록 만든 이유는 무엇인가요?

HamBeomJoon replied:
uiState를 도입하면서 liveData -> StateFlow로 변경했습니다.
LiveData는 observe() 시점에 Lifecycle을 자동 감지해서 UI 업데이트를 하는데, StateFlow는 그냥 콜드 스트림으로, collect 시점에 Lifecycle 고려가 필요합니다. XML 데이터바인딩은 collect를 직접 제어할 수 없다고 하네요.

`val liveData = viewModel.stateFlow.asLiveData()` LiveData 변환 후 바인딩하면 되긴하는데, 이러면 flow의 이점이 사라져서 Fragment/Activity에서 관리하는 방식을 채택했습니다. 권장되는 방식이기도 합니다

rosemin928 reviewed (approved):
수고많으셨습니다 미플~!
개인적으로 궁금한 점과, 몇 가지 제안을 코멘트에 적어두었으니 천천히 확인하고 반영 및 답변을 달아주세요! 😊
uiState 깔끔하네요 👍

+) 코드래빗의 리뷰도 한 번 확인해주세요!

HamBeomJoon committed:
[AN] refactor: HomeViewModel 로딩 로직 개선 #711

*   [AN] refactor: 중복된 로딩 및 예외 처리 로직 `safeLoad` 함수로 추출
*   [AN] refactor: `try-finally` 및 `NonCancellable` 컨텍스트를 사용하여 로딩 상태를 안정적으로 관리하도록 개선

HamBeomJoon committed:
[AN] refactor: 홈 화면 어댑터 패키지 구조 변경 #711

*   [AN] refactor: 홈 화면 관련 어댑터들을 `adapter` 하위 패키지로 이동

HamBeomJoon committed:
[AN] refactor: 홈 화면 로딩 키를 enum 클래스로 관리 #711

`HomeViewModel`에서 로딩 작업을 추적하기 위해 사용하던 문자열 키를 `HomeLoadKey`라는 enum 클래스로 대체하여 타입 안정성을 개선했습니다.

HamBeomJoon commented on android/app/src/main/java/com/onair/hearit/presentation/home/HomeFragment.kt:
```diff
@@ -170,50 +172,94 @@ class HomeFragment :
-        viewModel.recentUploadHearits.observe(viewLifecycleOwner) { recentUploadHearits ->
-            recentUploadAdapter.submitList(recentUploadHearits)
-        }
+    private fun updateRecentUploadSection(
+        recentUploadHearits: List<RecentUploadHearit>,
+        shouldShow: Boolean,
+    ) {
+        binding.tvHomeRecentUploadTitle.isVisible = shouldShow
+        recentUploadAdapter.submitList(recentUploadHearits)
+    }
```
최근 업로드된 히어릿은 비어있는 경우가 없을거라 생각해 처리안했습니다

HamBeomJoon committed:
[AN] refactor: loadingKeys를 UiState에서 관리하도록 변경 #711

*   [AN] refactor: `HomeUiState`의 로딩 상태를 Boolean에서 Set으로 변경
*   [AN] refactor: 여러 비동기 작업의 로딩 상태를 개별적으로 추적하도록 `HomeViewModel` 수정

HamBeomJoon committed:
[AN] chore: toast 메세지 정리 #711

HamBeomJoon committed:
[AN] refactor: 추천 카테고리 관련 용어 변경 및 로직 개선 #711

*   [AN] refactor: 추천 카테고리 관련 네이밍 변경 (grouped -> recommendation)
*   [AN] refactor: HomeViewModel의 safeLoad 함수 반환 타입 변경
